### PR TITLE
Add load balancing hostname_levenshtein_distance

### PR DESCRIPTION
**Pull request by JackyWoo**
### Usage

My ClickHouse Cluster is deployed in private clould. The host FQDN is like `rhbdp-ch-000449-clickhouse-0-0` in which the first `-0` represents shard number 0 and the second `-0` represents replica number 0.

If user request host `-0-0`, we want all hosts participating query are replica 0 in which way we implement resource isolation.  So we need a proper load balancer.

### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add load balancing test_hostname_levenshtein_distance.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit f81f0b6e3d96843205c1422cb0902f6e8bfb87c5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54826/f81f0b6e3d96843205c1422cb0902f6e8bfb87c5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Review by tavplubix — APPROVED** [review]

**Comment by den-crane** [comment]
Just add one more type of the balancing.
Otherwise it will break thousands of installations.

**Comment by den-crane** [comment]
I would add

nearestHostPrefix -- `eu-mych12333454.mydomain` <-> `ue-mychx.mydomain`
nearestHostPostfix --   `mych12333454.ue.mydomain` <-> `mychx.ue.mydomain`

**Comment by tavplubix** [comment]
Let's keep the old behavior for the "nearest hostname" type of load balancing and introduce a new type that uses Levenshtein distance

**Comment by JackyWoo** [comment]
@tavplubix add a new load balancing stretagy `hostname_levenshtein_distance` and it also works for zookeeper.

**Comment by JackyWoo** [comment]
@tavplubix CI was stucked by the following error:

```
D_LIBCPP_DEBUG=0 -std=c++23   -D OS_LINUX -Werror -nostdinc++ -mwaitpkg -MD -MT src/CMakeFiles/dbms.dir/Compression/CompressionCodecDeflateQpl.cpp.o -MF src/CMakeFiles/dbms.dir/Compression/CompressionCodecDeflateQpl.cpp.o.d -o src/CMakeFiles/dbms.dir/Compression/CompressionCodecDeflateQpl.cpp.o -c /build/src/Compression/CompressionCodecDeflateQpl.cpp
Oct 07 04:32:13 /build/src/Compression/CompressionCodecDeflateQpl.cpp:405:62: error: no member named 'getPageSize' in the global namespace; did you mean 'getpagesize'? [clang-diagnostic-error]
Oct 07 04:32:13   405 |     for (char * p = buffer; p < buffer + buffer_size; p += ::getPageSize()/(sizeof(*p)))
Oct 07 04:32:13       |                                                            ~~^~~~~~~~~~~
Oct 07 04:32:13       |                                                              getpagesize
Oct 07 04:32:13 /build/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc/usr/include/unistd.h:979:12: note: 'getpagesize' declared here
Oct 07 04:32:13   979 | extern int getpagesize (void)  __THROW __attribute__ ((__const__));
Oct 07 04:32:13       |            ^
Oct 07 04:32:13 23059 warnings and 1 error generated.
Oct 07 04:32:13 Error while processing /build/src/Compression/CompressionCodecDeflateQpl.cpp.
Oct 07 04:32:13 Suppressed 23293 warnings (23059 in non-user code, 234 NOLINT).
Oct 07 04:32:13 Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Oct 07 04:32:13 Found compiler error(s).
```

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
> CI was stucked by the following error

Probably it was broken in master, could you please merge fresh master into your branch?

**Comment by JackyWoo** [comment]
@tavplubix got same build error

File: src/Common/NamePrompter.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #pragma once
 
 #include <base/types.h>
-#include <Common/PODArray.h>
+#include <Common/levenshteinDistance.h>
```
Comment (tavplubix) [inline src/Common/NamePrompter.h:4]:
```suggestion
#include <Common/PODArray.h>
#include <Common/levenshteinDistance.h>
```

**Comment by tavplubix** [comment]
Integration tests (release) [2/4] - https://github.com/ClickHouse/ClickHouse/issues/52554
Stateless tests (tsan, s3 storage) [5/5] - data race in KQL
Performance Comparison looks flaky

**Pull request merged**